spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala

```diff
@@ -297,5 +301,86 @@ object SparkTableUtil {
       )
     }
   }
+
+  def buildManifest(table: Table,
+                    sparkDataFiles: Seq[SparkDataFile],
+                    partitionSpec: PartitionSpec): ManifestFile = {
+    val outputFile = table.io
+      .newOutputFile(FileFormat.AVRO.addExtension("/tmp/" + UUID.randomUUID.toString))
+    val writer = ManifestWriter.write(partitionSpec, outputFile)
+    try {
+      for (file <- sparkDataFiles) {
+        writer.add(file.toDataFile(partitionSpec))
+      }
+    } finally {
+      writer.close()
+    }
+
+    writer.toManifestFile
+  }
+
+  def partitionToMap(partition: String): Map[String, String] = {
+    val map = new mutable.HashMap[String, String]()
+    val list = partition.split("/")
+    list.foreach { str =>
+      val kv = str.split("=")
+      map.put(kv(0), kv(1))
+    }
+
+    map.toMap
+  }
+
+  /**
+   * Migrate a spark table to a iceberg table.
+   *
+   * The migration uses the spark session to get table metadata.
+   *
+   * @param dbName the database name of the table to be migrated
+   * @param tableName the table to be migrated
+   *
+   * @return table the migrated table
+   */
+  def migrateSparkTable(dbName: String, tableName: String, table: Table): Table = {
+    val sparkSession = SparkSession.builder().getOrCreate()
+
+    if (!sparkSession.catalog.tableExists(dbName, tableName)) {
+      throw new NoSuchTableException(s"Table $dbName.$tableName does not exist")
+    }
+
+    val tableMetadata = sparkSession.sessionState.catalog.
+      getTableMetadata(new TableIdentifier(tableName, Some(dbName)))
+
+    val format = tableMetadata.provider.getOrElse("none")
+    if (format != "avro" && format != "parquet" && format != "orc") {
+      throw new UnsupportedOperationException(s"Unknown partition format: $format")
+    }
+
+    val location = tableMetadata.location.toString
+    val partitionSpec = SparkSchemaUtil.specForTable(sparkSession, s"$dbName.$tableName")
+
+    val fastAppender = table.newFastAppend()
+
+    val partitions = sparkSession.sessionState.catalog.externalCatalog
+      .listPartitionNames(dbName, tableName)
+    if (partitions.isEmpty) {
+      val dataFiles = SparkTableUtil.listPartition(Map.empty[String, String], location, format)
+      fastAppender.appendManifest(buildManifest(table, dataFiles, PartitionSpec.unpartitioned))
+    } else {
+      // Retrieve data files according to partition. result = [[datafiles], [datafiles]]
+      val dataFiles = partitions.map { e =>
+        SparkTableUtil.listPartition(partitionToMap(e), location + "/" + e, format)
+      }
+
+      // Append manifest for each partition
+      dataFiles.foreach { partition =>
+        fastAppender.appendManifest(buildManifest(table, partition, partitionSpec))
+      }
+
+      fastAppender.apply()
```

Comment:
apply(TableMetadata base) in FastAppend

Reply:
Limitation: Single thread update

Reply:
Thanks, I moved this to https://github.com/apache/incubator-iceberg/pull/374.